File: ruby_practice_problems/02_netflix/group_anagrams.rb

def group_anagrams(strs)
  encoded = {}
  strs.each do |str|
    code = [0] * 26
    str.each_char do |ch|
      idx = ch.ord - 'a'.ord
      code[idx] += 1
    end
    dict_code = code.join("")
    if encoded[dict_code] != nil
      encoded[dict_code].push(str)
    else
      encoded[dict_code] = [str]
    end
  end

  result = []
  encoded.keys().each do |key|
    result.push(encoded[key])
  end
  return result
end

File: python_practice_problems/22_what_to_track/group_anagrams.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Wed Apr 26 17:43:20 2023

@author: johnmorgan
"""

# O(n * k) - word * letters/word

def group_anagrams(strs):
    str_dict = {}
    for astr in strs:
        chrs = [0] * 26
        for letter in astr:
            val = ord(letter) - 97
            chrs[val] += 1
        for index in range(len(chrs)):
            chrs[index] = str(chrs[index])
        key = "".join(chrs)
        if key in str_dict:
            str_dict[key] += [astr]
        else:
            str_dict[key] = [astr]
    result = []
    for key in str_dict:
        result.append(str_dict[key])
    return result